File: src/plugins/concerts.ts
import { IMatrixClient, Plugin, CommandDef, MessageContext } from "./base";
import { RateLimitsPlugin } from "./ratelimits";
import { getDb } from "../db";
import logger from "../utils/logger";

interface BandsintownEvent {
  venue: { name: string; city: string; country: string };
  datetime: string;
  offers: any[];
  lineup: string[];
}

export class ConcertsPlugin extends Plugin {
  private rateLimitPlugin: RateLimitsPlugin;

  constructor(client: IMatrixClient, rateLimitPlugin: RateLimitsPlugin) {
    super(client);
    this.rateLimitPlugin = rateLimitPlugin;
  }

  get name() {
    return "concerts";
  }

  get commands(): CommandDef[] {
    return [
      {
        name: "concerts",
        description: "Search upcoming concerts or manage your artist watchlist",
        usage: "!concerts <artist> | !concerts watch <artist> | !concerts watching | !concerts unwatch <artist>",
      },
    ];
  }

  async onMessage(ctx: MessageContext): Promise<void> {
    if (!this.isCommand(ctx.body, "concerts")) return;

    const args = this.getArgs(ctx.body, "concerts").trim();

    if (!args) {
      await this.sendReply(
        ctx.roomId,
        ctx.eventId,
        "Usage: !concerts <artist> | !concerts watch <artist> | !concerts watching | !concerts unwatch <artist>"
      );
      return;
    }

    if (args.toLowerCase().startsWith("watch ")) {
      const artist = args.slice(6).trim();
      if (!artist) {
        await this.sendReply(ctx.roomId, ctx.eventId, "Specify an artist to watch.");
        return;
      }
      await this.handleWatch(ctx, artist);
      return;
    }

    if (args.toLowerCase() === "watching") {
      await this.handleWatching(ctx);
      return;
    }

    if (args.toLowerCase().startsWith("unwatch ")) {
      const artist = args.slice(8).trim();
      if (!artist) {
        await this.sendReply(ctx.roomId, ctx.eventId, "Specify an artist to unwatch.");
        return;
      }
      await this.handleUnwatch(ctx, artist);
      return;
    }

    // Default: search by artist name
    await this.handleArtistSearch(ctx, args);
  }

  private getApiKey(): string | undefined {
    return process.env.BANDSINTOWN_API_KEY;
  }

  private async handleArtistSearch(ctx: MessageContext, artist: string): Promise<void> {
    const key = this.getApiKey();
    if (!key) {
      await this.sendReply(ctx.roomId, ctx.eventId, "Bandsintown API key not configured.");
      return;
    }

    const allowed = this.rateLimitPlugin.checkLimit(
      ctx.sender,
      "concerts",
      parseInt(process.env.RATELIMIT_CONCERTS ?? "10")
    );
    if (!allowed) {
      await this.sendReply(ctx.roomId, ctx.eventId, "You've reached your daily concerts lookup quota. Try again tomorrow.");
      return;
    }

    try {
      const events = await this.fetchArtistEvents(artist, key);

      if (!events || events.length === 0) {
        await this.sendReply(ctx.roomId, ctx.eventId, `No upcoming events found for "${artist}".`);
        return;
      }

      const lines = events.slice(0, 10).map((evt) => {
        const lineup = evt.lineup?.join(", ") ?? artist;
        const venue = evt.venue?.name ?? "Unknown Venue";
        const city = evt.venue?.city ?? "Unknown City";
        const date = this.formatDate(evt.datetime);
        return `${lineup} — ${venue}, ${city} — ${date}`;
      });

      const header = `Upcoming events for "${artist}" (${Math.min(events.length, 10)} of ${events.length}):`;
      await this.sendMessage(ctx.roomId, `${header}\n${lines.join("\n")}`);
    } catch (err) {
      logger.error(`Concerts artist search failed for "${artist}": ${err}`);
      await this.sendReply(ctx.roomId, ctx.eventId, "Failed to fetch concert data. Try again later.");
    }
  }

  private async fetchArtistEvents(artist: string, apiKey: string): Promise<BandsintownEvent[]> {
    const cacheKey = artist.toLowerCase();
    const db = getDb();

    // Check cache (6 hours = 21600 seconds)
    const cached = db
      .prepare(`SELECT data FROM concerts_cache WHERE location_key = ? AND cached_at > unixepoch() - 21600`)
      .get(cacheKey) as { data: string } | undefined;

    if (cached) {
      return JSON.parse(cached.data);
    }

    const url = `https://rest.bandsintown.com/artists/${encodeURIComponent(artist)}/events?app_id=${encodeURIComponent(apiKey)}&date=upcoming`;
    const res = await fetch(url);

    if (!res.ok) {
      logger.warn(`Bandsintown API returned ${res.status} for artist "${artist}"`);
      return [];
    }

    const data = (await res.json()) as BandsintownEvent[];

    if (!Array.isArray(data)) {
      return [];
    }

    // Cache the result
    db.prepare(
      `INSERT INTO concerts_cache (location_key, data, cached_at) VALUES (?, ?, unixepoch())
       ON CONFLICT(location_key) DO UPDATE SET data = excluded.data, cached_at = unixepoch()`
    ).run(cacheKey, JSON.stringify(data));

    return data;
  }

  private async handleWatch(ctx: MessageContext, artist: string): Promise<void> {
    const db = getDb();

    try {
      db.prepare(
        `INSERT INTO concert_watchlist (user_id, room_id, artist, created_at) VALUES (?, ?, ?, unixepoch())`
      ).run(ctx.sender, ctx.roomId, artist.toLowerCase());

      await this.sendReply(ctx.roomId, ctx.eventId, `Now watching "${artist}" for upcoming concerts.`);
    } catch (err: any) {
      if (err?.code === "SQLITE_CONSTRAINT_UNIQUE" || err?.message?.includes("UNIQUE")) {
        await this.sendReply(ctx.roomId, ctx.eventId, `You're already watching "${artist}".`);
      } else {
        logger.error(`Failed to add concert watch for ${ctx.sender}: ${err}`);
        await this.sendReply(ctx.roomId, ctx.eventId, "Failed to add artist to watchlist.");
      }
    }
  }

  private async handleWatching(ctx: MessageContext): Promise<void> {
    const db = getDb();

    const rows = db
      .prepare(`SELECT artist FROM concert_watchlist WHERE user_id = ? AND room_id = ? ORDER BY artist`)
      .all(ctx.sender, ctx.roomId) as { artist: string }[];

    if (rows.length === 0) {
      await this.sendReply(ctx.roomId, ctx.eventId, "You're not watching any artists. Use !concerts watch <artist> to add one.");
      return;
    }

    const list = rows.map((r) => `- ${r.artist}`).join("\n");
    await this.sendMessage(ctx.roomId, `Your watched artists:\n${list}`);
  }

  private async handleUnwatch(ctx: MessageContext, artist: string): Promise<void> {
    const db = getDb();

    const result = db
      .prepare(`DELETE FROM concert_watchlist WHERE user_id = ? AND room_id = ? AND artist = ?`)
      .run(ctx.sender, ctx.roomId, artist.toLowerCase());

    if (result.changes > 0) {
      await this.sendReply(ctx.roomId, ctx.eventId, `Stopped watching "${artist}".`);
    } else {
      await this.sendReply(ctx.roomId, ctx.eventId, `You weren't watching "${artist}".`);
    }
  }

  /**
   * Called by the scheduler to post a weekly concert digest.
   * Only posts on Sundays (day 0).
   */
  async postWeeklyDigest(botRooms: string[]): Promise<void> {
    const today = new Date();
    if (today.getUTCDay() !== 0) return;

    const key = this.getApiKey();
    if (!key) {
      logger.warn("BANDSINTOWN_API_KEY not set, skipping concert digest");
      return;
    }

    const db = getDb();

    // Get all unique watched artists across all users
    const watchedArtists = db
      .prepare(`SELECT DISTINCT artist FROM concert_watchlist`)
      .all() as { artist: string }[];

    if (watchedArtists.length === 0) return;

    // Calculate the date range for the next 7 days
    const now = new Date();
    const weekFromNow = new Date(now);
    weekFromNow.setUTCDate(weekFromNow.getUTCDate() + 7);

    const startIso = now.toISOString().slice(0, 10);
    const endIso = weekFromNow.toISOString().slice(0, 10);

    // Fetch events for each watched artist
    const upcomingShows: { artist: string; events: BandsintownEvent[] }[] = [];

    for (const { artist } of watchedArtists) {
      try {
        const events = await this.fetchArtistEvents(artist, key);
        const thisWeek = events.filter((evt) => {
          const eventDate = evt.datetime?.slice(0, 10);
          return eventDate && eventDate >= startIso && eventDate <= endIso;
        });

        if (thisWeek.length > 0) {
          upcomingShows.push({ artist, events: thisWeek });
        }
      } catch (err) {
        logger.error(`Failed to fetch events for watched artist "${artist}": ${err}`);
      }
    }

    if (upcomingShows.length === 0) return;

    // Build summary message for bot rooms
    const lines: string[] = ["Concert Digest — This Week:", ""];
    for (const { artist, events } of upcomingShows) {
      for (const evt of events) {
        const venue = evt.venue?.name ?? "Unknown Venue";
        const city = evt.venue?.city ?? "Unknown City";
        const date = this.formatDate(evt.datetime);
        const lineup = evt.lineup?.join(", ") ?? artist;
        lines.push(`${lineup} — ${venue}, ${city} — ${date}`);
      }
    }

    const message = lines.join("\n");

    for (const roomId of botRooms) {
      try {
        await this.sendMessage(roomId, message);
      } catch (err) {
        logger.error(`Failed to post concert digest to ${roomId}: ${err}`);
      }
    }

    // DM users who are watching artists with shows this week
    const artistsWithShows = new Set(upcomingShows.map((s) => s.artist));

    const watchers = db
      .prepare(`SELECT DISTINCT user_id, artist FROM concert_watchlist WHERE artist IN (${[...artistsWithShows].map(() => "?").join(",")})`)
      .all(...artistsWithShows) as { user_id: string; artist: string }[];

    // Group by user
    const userArtists = new Map<string, string[]>();
    for (const { user_id, artist } of watchers) {
      if (!userArtists.has(user_id)) userArtists.set(user_id, []);
      userArtists.get(user_id)!.push(artist);
    }

    for (const [userId, artists] of userArtists) {
      const dmLines: string[] = ["Concerts this week for your watched artists:", ""];
      for (const artist of artists) {
        const showData = upcomingShows.find((s) => s.artist === artist);
        if (!showData) continue;
        for (const evt of showData.events) {
          const venue = evt.venue?.name ?? "Unknown Venue";
          const city = evt.venue?.city ?? "Unknown City";
          const date = this.formatDate(evt.datetime);
          const lineup = evt.lineup?.join(", ") ?? artist;
          dmLines.push(`${lineup} — ${venue}, ${city} — ${date}`);
        }
      }

      try {
        await this.sendDm(userId, dmLines.join("\n"));
      } catch (err) {
        logger.error(`Failed to DM concert digest to ${userId}: ${err}`);
      }
    }
  }

  private formatDate(datetime: string): string {
    try {
      const d = new Date(datetime);
      return d.toLocaleDateString("en-US", {
        weekday: "short",
        month: "short",
        day: "numeric",
        year: "numeric",
        timeZone: "UTC",
      });
    } catch {
      return datetime?.slice(0, 10) ?? "Unknown Date";
    }
  }
}

File: internal/plugin/concerts.go
package plugin

import (
	"context"
	"encoding/json"
	"fmt"
	"log/slog"
	"net/http"
	"net/url"
	"os"
	"strings"
	"sync"
	"time"

	"gogobee/internal/db"

	"maunium.net/go/mautrix"
	"maunium.net/go/mautrix/id"
)

// bandsintownEvent represents an event from the Bandsintown API.
type bandsintownEvent struct {
	ID       string `json:"id"`
	URL      string `json:"url"`
	DateTime string `json:"datetime"`
	Title    string `json:"title"`
	Venue    struct {
		Name      string `json:"name"`
		City      string `json:"city"`
		Region    string `json:"region"`
		Country   string `json:"country"`
		Latitude  string `json:"latitude"`
		Longitude string `json:"longitude"`
	} `json:"venue"`
	Lineup    []string `json:"lineup"`
	OnSaleAt  string   `json:"on_sale_datetime"`
	Offers    []struct {
		Type   string `json:"type"`
		URL    string `json:"url"`
		Status string `json:"status"`
	} `json:"offers"`
}

// ConcertsPlugin provides concert lookups via Bandsintown.
type ConcertsPlugin struct {
	Base
	apiKey      string
	httpClient  *http.Client
	rateLimiter *RateLimitsPlugin
	mu          sync.Mutex
	cooldowns   map[string]time.Time // keyed by userID+artist
}

// NewConcertsPlugin creates a new ConcertsPlugin.
func NewConcertsPlugin(client *mautrix.Client, rateLimiter *RateLimitsPlugin) *ConcertsPlugin {
	return &ConcertsPlugin{
		Base:        NewBase(client),
		apiKey:      os.Getenv("BANDSINTOWN_API_KEY"),
		rateLimiter: rateLimiter,
		httpClient: &http.Client{
			Timeout: 10 * time.Second,
		},
		cooldowns: make(map[string]time.Time),
	}
}

func (p *ConcertsPlugin) Name() string { return "concerts" }

func (p *ConcertsPlugin) Commands() []CommandDef {
	return []CommandDef{
		{Name: "concerts", Description: "Search upcoming concerts for an artist", Usage: "!concerts <artist>", Category: "Entertainment"},
		{Name: "concerts watch", Description: "Watch an artist for concert updates", Usage: "!concerts watch <artist>", Category: "Entertainment"},
		{Name: "concerts watching", Description: "List your watched artists", Usage: "!concerts watching", Category: "Entertainment"},
		{Name: "concerts unwatch", Description: "Stop watching an artist", Usage: "!concerts unwatch <artist>", Category: "Entertainment"},
	}
}

func (p *ConcertsPlugin) Init() error { return nil }

func (p *ConcertsPlugin) OnReaction(_ ReactionContext) error { return nil }

func (p *ConcertsPlugin) OnMessage(ctx MessageContext) error {
	if !p.IsCommand(ctx.Body, "concerts") {
		return nil
	}

	args := p.GetArgs(ctx.Body, "concerts")
	if args == "" {
		return p.SendReply(ctx.RoomID, ctx.EventID, "Usage: !concerts <artist> | !concerts watch|watching|unwatch <artist>")
	}

	parts := strings.SplitN(args, " ", 2)
	sub := strings.ToLower(parts[0])

	switch sub {
	case "watch":
		if len(parts) < 2 || strings.TrimSpace(parts[1]) == "" {
			return p.SendReply(ctx.RoomID, ctx.EventID, "Usage: !concerts watch <artist>")
		}
		return p.handleWatch(ctx, strings.TrimSpace(parts[1]))
	case "watching":
		return p.handleWatching(ctx)
	case "unwatch":
		if len(parts) < 2 || strings.TrimSpace(parts[1]) == "" {
			return p.SendReply(ctx.RoomID, ctx.EventID, "Usage: !concerts unwatch <artist>")
		}
		return p.handleUnwatch(ctx, strings.TrimSpace(parts[1]))
	default:
		return p.handleSearch(ctx, args)
	}
}

func (p *ConcertsPlugin) handleSearch(ctx MessageContext, artist string) error {
	if p.apiKey == "" {
		return p.SendReply(ctx.RoomID, ctx.EventID, "Concert lookups are not configured (missing API key).")
	}

	// Daily rate limit (configurable, default 10/day to match TS version)
	concertLimit := 10
	if v := os.Getenv("RATELIMIT_CONCERTS"); v != "" {
		if n, err := fmt.Sscanf(v, "%d", &concertLimit); n != 1 || err != nil {
			concertLimit = 10
		}
	}
	if p.rateLimiter != nil && !p.rateLimiter.CheckLimit(ctx.Sender, "concerts", concertLimit) {
		return p.SendReply(ctx.RoomID, ctx.EventID, "Concert search rate limit reached for today.")
	}

	// Per-search cooldown
	cooldownKey := string(ctx.Sender) + ":" + strings.ToLower(artist)
	p.mu.Lock()
	if last, ok := p.cooldowns[cooldownKey]; ok && time.Since(last) < 30*time.Second {
		p.mu.Unlock()
		return p.SendReply(ctx.RoomID, ctx.EventID, "Please wait a moment before searching for this artist again.")
	}
	p.cooldowns[cooldownKey] = time.Now()
	p.mu.Unlock()

	events, err := p.fetchEvents(artist)
	if err != nil {
		slog.Error("concerts: fetch failed", "artist", artist, "err", err)
		return p.SendReply(ctx.RoomID, ctx.EventID, "Failed to fetch concert data.")
	}

	if len(events) == 0 {
		return p.SendReply(ctx.RoomID, ctx.EventID, fmt.Sprintf("No upcoming concerts found for %s.", artist))
	}

	// Show up to 5 events
	limit := 5
	if len(events) < limit {
		limit = len(events)
	}

	var sb strings.Builder
	sb.WriteString(fmt.Sprintf("Upcoming concerts for %s:\n\n", artist))

	for i := 0; i < limit; i++ {
		ev := events[i]
		dateStr := p.formatEventDate(ev.DateTime)
		location := p.formatLocation(ev.Venue.City, ev.Venue.Region, ev.Venue.Country)
		sb.WriteString(fmt.Sprintf("%d. %s\n", i+1, dateStr))
		sb.WriteString(fmt.Sprintf("   Venue: %s\n", ev.Venue.Name))
		sb.WriteString(fmt.Sprintf("   Location: %s\n", location))
		if ev.URL != "" {
			sb.WriteString(fmt.Sprintf("   Tickets: %s\n", ev.URL))
		}
		if i < limit-1 {
			sb.WriteString("\n")
		}
	}

	if len(events) > limit {
		sb.WriteString(fmt.Sprintf("\n...and %d more events.", len(events)-limit))
	}

	return p.SendReply(ctx.RoomID, ctx.EventID, sb.String())
}

func (p *ConcertsPlugin) fetchEvents(artist string) ([]bandsintownEvent, error) {
	d := db.Get()
	artistKey := strings.ToLower(artist)

	// Check cache (6h TTL)
	var cached string
	var cachedAt int64
	err := d.QueryRow(
		`SELECT data, cached_at FROM concerts_cache WHERE artist = ?`, artistKey,
	).Scan(&cached, &cachedAt)
	if err == nil && time.Now().Unix()-cachedAt < 6*3600 {
		var events []bandsintownEvent
		if json.Unmarshal([]byte(cached), &events) == nil {
			return events, nil
		}
	}

	// Fetch from API
	encoded := url.PathEscape(artist)
	apiURL := fmt.Sprintf("https://rest.bandsintown.com/artists/%s/events?app_id=%s&date=upcoming",
		encoded, p.apiKey)

	req, err := http.NewRequestWithContext(context.Background(), http.MethodGet, apiURL, nil)
	if err != nil {
		return nil, err
	}
	req.Header.Set("Accept", "application/json")

	resp, err := p.httpClient.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("bandsintown returned status %d", resp.StatusCode)
	}

	var events []bandsintownEvent
	if err := json.NewDecoder(resp.Body).Decode(&events); err != nil {
		return nil, err
	}

	// Update cache
	data, _ := json.Marshal(events)
	_, err = d.Exec(
		`INSERT INTO concerts_cache (artist, data, cached_at) VALUES (?, ?, ?)
		 ON CONFLICT(artist) DO UPDATE SET data = ?, cached_at = ?`,
		artistKey, string(data), time.Now().Unix(), string(data), time.Now().Unix(),
	)
	if err != nil {
		slog.Error("concerts: cache write", "err", err)
	}

	return events, nil
}

func (p *ConcertsPlugin) formatEventDate(dateStr string) string {
	layouts := []string{
		"2006-01-02T15:04:05",
		time.RFC3339,
		"2006-01-02",
	}
	for _, layout := range layouts {
		if t, err := time.Parse(layout, dateStr); err == nil {
			return t.Format("Mon, Jan 2, 2006 3:04 PM")
		}
	}
	return dateStr
}

func (p *ConcertsPlugin) formatLocation(city, region, country string) string {
	parts := []string{}
	if city != "" {
		parts = append(parts, city)
	}
	if region != "" {
		parts = append(parts, region)
	}
	if country != "" {
		parts = append(parts, country)
	}
	return strings.Join(parts, ", ")
}

func (p *ConcertsPlugin) handleWatch(ctx MessageContext, artist string) error {
	d := db.Get()
	artistKey := strings.ToLower(artist)
	_, err := d.Exec(
		`INSERT INTO concert_watchlist (user_id, artist, room_id) VALUES (?, ?, ?)
		 ON CONFLICT(user_id, artist) DO NOTHING`,
		string(ctx.Sender), artistKey, string(ctx.RoomID),
	)
	if err != nil {
		slog.Error("concerts: watch add", "err", err)
		return p.SendReply(ctx.RoomID, ctx.EventID, "Failed to add artist to watchlist.")
	}
	return p.SendReply(ctx.RoomID, ctx.EventID, fmt.Sprintf("Now watching %s for concert updates.", artist))
}

func (p *ConcertsPlugin) handleWatching(ctx MessageContext) error {
	d := db.Get()
	rows, err := d.Query(
		`SELECT artist FROM concert_watchlist WHERE user_id = ? ORDER BY artist`,
		string(ctx.Sender),
	)
	if err != nil {
		slog.Error("concerts: watching list", "err", err)
		return p.SendReply(ctx.RoomID, ctx.EventID, "Failed to load watchlist.")
	}
	defer rows.Close()

	var artists []string
	for rows.Next() {
		var a string
		if err := rows.Scan(&a); err != nil {
			continue
		}
		artists = append(artists, a)
	}

	if len(artists) == 0 {
		return p.SendReply(ctx.RoomID, ctx.EventID, "You're not watching any artists. Use !concerts watch <artist> to start.")
	}

	var sb strings.Builder
	sb.WriteString("Your Concert Watchlist:\n")
	for _, a := range artists {
		sb.WriteString(fmt.Sprintf("  - %s\n", a))
	}
	return p.SendReply(ctx.RoomID, ctx.EventID, sb.String())
}

func (p *ConcertsPlugin) handleUnwatch(ctx MessageContext, artist string) error {
	d := db.Get()
	artistKey := strings.ToLower(artist)
	res, err := d.Exec(
		`DELETE FROM concert_watchlist WHERE user_id = ? AND artist = ?`,
		string(ctx.Sender), artistKey,
	)
	if err != nil {
		slog.Error("concerts: unwatch", "err", err)
		return p.SendReply(ctx.RoomID, ctx.EventID, "Failed to remove artist from watchlist.")
	}
	n, _ := res.RowsAffected()
	if n == 0 {
		return p.SendReply(ctx.RoomID, ctx.EventID, fmt.Sprintf("%s is not in your watchlist.", artist))
	}
	return p.SendReply(ctx.RoomID, ctx.EventID, fmt.Sprintf("Stopped watching %s.", artist))
}

// PostWeeklyDigest sends a weekly concert digest to a room and DMs watchers about upcoming shows.
// Intended to be called by the scheduler on Sundays.
func (p *ConcertsPlugin) PostWeeklyDigest(roomID id.RoomID) {
	if p.apiKey == "" {
		slog.Warn("concerts: skipping weekly digest, no API key configured")
		return
	}

	year, week := time.Now().UTC().ISOWeek()
	weekKey := fmt.Sprintf("%d-W%02d", year, week)
	if db.JobCompleted("concert_digest", weekKey) {
		slog.Info("concerts: already sent digest this week", "week", weekKey)
		return
	}
	// Mark completed at the end only if we succeed (not deferred)
	success := false
	defer func() {
		if success {
			db.MarkJobCompleted("concert_digest", weekKey)
		}
	}()

	d := db.Get()

	// Get all unique watched artists
	rows, err := d.Query(`SELECT DISTINCT artist FROM concert_watchlist`)
	if err != nil {
		slog.Error("concerts: weekly digest query", "err", err)
		return
	}

	type watcherInfo struct {
		userID id.UserID
		roomID id.RoomID
	}

	artistWatchers := make(map[string][]watcherInfo)
	var allArtists []string

	// First collect all artists
	for rows.Next() {
		var artist string
		if err := rows.Scan(&artist); err != nil {
			continue
		}
		allArtists = append(allArtists, artist)
	}
	rows.Close()

	// Get watchers per artist
	for _, artist := range allArtists {
		wRows, err := d.Query(
			`SELECT user_id, room_id FROM concert_watchlist WHERE artist = ?`, artist,
		)
		if err != nil {
			continue
		}
		for wRows.Next() {
			var uid, rid string
			if err := wRows.Scan(&uid, &rid); err != nil {
				continue
			}
			artistWatchers[artist] = append(artistWatchers[artist], watcherInfo{
				userID: id.UserID(uid),
				roomID: id.RoomID(rid),
			})
		}
		wRows.Close()
	}

	// For each artist, fetch events and notify watchers
	for artist, watchers := range artistWatchers {
		events, err := p.fetchEvents(artist)
		if err != nil {
			slog.Error("concerts: weekly digest fetch", "artist", artist, "err", err)
			continue
		}

		if len(events) == 0 {
			continue
		}

		// Filter to events in the next 2 weeks
		now := time.Now()
		twoWeeks := now.Add(14 * 24 * time.Hour)
		var upcoming []bandsintownEvent
		for _, ev := range events {
			t, err := time.Parse("2006-01-02T15:04:05", ev.DateTime)
			if err != nil {
				t, err = time.Parse(time.RFC3339, ev.DateTime)
			}
			if err == nil && t.After(now) && t.Before(twoWeeks) {
				upcoming = append(upcoming, ev)
			}
		}

		if len(upcoming) == 0 {
			continue
		}

		// Build message
		var sb strings.Builder
		sb.WriteString(fmt.Sprintf("Upcoming shows for %s (next 2 weeks):\n\n", artist))
		for i, ev := range upcoming {
			if i >= 5 {
				sb.WriteString(fmt.Sprintf("\n...and %d more.", len(upcoming)-5))
				break
			}
			dateStr := p.formatEventDate(ev.DateTime)
			location := p.formatLocation(ev.Venue.City, ev.Venue.Region, ev.Venue.Country)
			sb.WriteString(fmt.Sprintf("- %s @ %s (%s)\n", dateStr, ev.Venue.Name, location))
		}

		msg := sb.String()
		for _, w := range watchers {
			if err := p.SendDM(w.userID, msg); err != nil {
				slog.Error("concerts: DM watcher", "user", w.userID, "err", err)
			}
		}
	}

	success = true
	slog.Info("concerts: weekly digest completed")
}
